pkg/agent/lease_counter.go

```diff
@@ -0,0 +1,81 @@
+package agent
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/agent/metrics"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/util"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coordinationv1 "k8s.io/client-go/kubernetes/typed/coordination/v1"
+)
+
+// A ServerLeaseCounter counts leases in the k8s apiserver to determine the
+// current proxy server count.
+type ServerLeaseCounter struct {
+	leaseClient   coordinationv1.LeaseInterface
+	selector      labels.Selector
+	fallbackCount int
+}
+
+// NewServerLeaseCounter creates a server counter that counts valid leases that match the label
+// selector and provides the fallback count (initially 0) if this fails.
+func NewServerLeaseCounter(k8sClient kubernetes.Interface, labelSelector labels.Selector) *ServerLeaseCounter {
+	return &ServerLeaseCounter{
+		leaseClient:   k8sClient.CoordinationV1().Leases(""),
+		selector:      labelSelector,
+		fallbackCount: 0,
+	}
+}
+
+// Count counts the number of leases in the apiserver matching the provided
+// label selector.
+//
+// In the event that no valid leases are found or lease listing fails, the
+// fallback count is returned. This fallback count is updated upon successful
+// discovery of valid leases.
+func (lc *ServerLeaseCounter) Count(ctx context.Context) int {
+	// Since the number of proxy servers is generally small (1-10), we opted against
+	// using a LIST and WATCH pattern and instead list all leases on each call.
```

Comment:
I think it's still worth considering using an informer here. It simplifies the caching, and if you plumb through the events could let the agents be more responsive to server changes.

Reply:
Added a comment about this.

Reply:
I see that one of the previous versions of this PR did use informer. Did you have trouble getting it to work? (from a quick look - is it possible that that version was missing a call to Start()? )

https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/643/commits/7a6b3f9d13fcdfe14de674742d5ffec42a7bf29e

I think Informer is preferrable even with a Lister-based approach (so: the logic looks like periodic listing, but we're hitting the cache instead of apiserver).

Reply:
@avrittrohwer asked me to switch it over to a direct list call in order to make observability into the success rate of the calls easier. I couldn't figure out a good way to hook up the Informer to publish metrics. It seems like people would generally prefer the Informer method, I can do this if it is desired!

Reply:
Will switch over to an Informer and hook in metrics in the Informer per some suggestions @avrittrohwer made!

Reply:
Done.